runtime/libia2/init.c

```diff
@@ -97,3 +100,134 @@ __attribute__((__noreturn__)) void ia2_reinit_stack_err(int i) {
          i, gettid());
   exit(1);
 }
+
+static void replace_suffix(char *name, int compartment) {
+    name[strlen(name) - 1] = '0' + compartment;
+}
+
+/* This function is expected to be called from the user-defined ia2_main once per compartment */
+static void ia2_protect_memory(const char *libs, int compartment, const char *extra_libraries) {
+    printf("%s: protecting %s with pkey %d\n", __func__, libs, compartment);
+
+    void *handle = RTLD_DEFAULT;
+    const char *dso_sym = "main";
+
+    if (strcmp(libs, "main")) {
+        void *handle = dlopen(libs, RTLD_GLOBAL | RTLD_NOW);
+        if (!handle) {
+            printf("%s: failed to dlopen %s for compartment %d\n", __func__, libs, compartment);
+            // TODO: use actual error codes
+            exit(-1);
+        }
+        // TODO: use a symbol that will always be defined. This is specific to tests/two_keys_minimal/plugin.c
+        dso_sym = "start_plugin";
+    }
+
+    void *dso_addr = dlsym(handle, dso_sym);
```

Comment:
Could `LD_PRELOAD` give you the wrong symbol here (and everywhere you call `dlsym`)?

Reply:
It might make sense to make all these symbols at least `protected` visibility, so they're harder to override.